runtime/eventring_consumer.c

```diff
@@ -0,0 +1,686 @@
+                                            int64_t timestamp,
+                                            ev_lifecycle lifecycle,
+                                            int64_t data)) {
+  cursor->lifecycle = f;
+}
+
+void caml_eventring_set_lost_events(struct caml_eventring_cursor *cursor,
+                                    int (*f)(int domain_id,
+                                              void *callback_data,
+                                              int lost_words)) {
+  cursor->lost_events = f;
+}
+
+/* frees a cursor obtained from caml_eventring_reader_create */
+void caml_eventring_free_cursor(struct caml_eventring_cursor *cursor) {
+  if (cursor->cursor_open) {
+    cursor->cursor_open = 0;
+#ifdef _WIN32
+    UnmapViewOfFile(cursor->metadata);
+    CloseHandle(cursor->ring_file_handle);
+    CloseHandle(cursor->ring_handle);
+#else
+    munmap(cursor->metadata, cursor->ring_file_size_bytes);
+#endif
+    caml_stat_free(cursor->current_positions);
+    caml_stat_free(cursor);
+  }
+}
+
+eventring_error
+caml_eventring_read_poll(struct caml_eventring_cursor *cursor,
+                         void *callback_data, uintnat max_events,
+                         uintnat *events_consumed) {
+  int consumed = 0;
+  int start_domain = cursor->next_read_domain;
+  uint64_t ring_head, ring_tail;
+  int early_exit = 0;
+
+  if (!cursor->cursor_open) {
+    return E_CURSOR_NOT_OPEN;
+  }
+
+  /* this loop looks a bit odd because we're iterating from the last domain
+     that we read from on the last read_poll call and then looping around.
+     This is necessary because in the case where the consumer can't keep up
+     with message production (i.e max_events is hit each time) it ensures that
+     messages are read from all domains, rather than just the first. */
+  for (int i = 0; i < cursor->metadata->max_domains && !early_exit; i++) {
+    int domain_num = (start_domain + i) % cursor->metadata->max_domains;
+
+    struct eventring_buffer_header *eventring_buffer_header =
+        (struct eventring_buffer_header *)(
+          (char*)cursor->metadata +
+          cursor->metadata->headers_offset +
+          domain_num * cursor->metadata->ring_header_size_bytes
+        );
+
+    uint64_t *ring_ptr = (uint64_t *)((char*)cursor->metadata +
+                                      cursor->metadata->data_offset +
+                                domain_num * cursor->metadata->ring_size_bytes);
+
+    do {
+      uint64_t buf[EVENTRING_MAX_MSG_LENGTH];
+      uint64_t ring_mask, header, msg_length;
+      ring_head = atomic_load_explicit(&eventring_buffer_header->ring_head,
+                                       memory_order_acquire);
+      ring_tail = atomic_load_explicit(&eventring_buffer_header->ring_tail,
+                                       memory_order_acquire);
+
+      if (ring_head > cursor->current_positions[domain_num]) {
+        if (cursor->lost_events) {
+          cursor->lost_events(domain_num, callback_data,
```

Comment:
Return value of lost_events (to trigger an early exit) is ignored here.

Reply:
Thanks, good spot.